## **PROJECT TITLE: PLAYSTORE APPS ANALYSIS**                                                             
### **BY TEAM DATA CRAWLERS**
TEAM MEMBERS:


*   G. K. JENNIFER 
*   AMUL GAURAV
*   IGBOKWE UCHANMA ADEOLA
                                                                                                 
                                                                                                             DATE: JANUARY 2023

**INTRODUCTION**

This was carried out as part of the HiCounselor Live Project. The datasets used comprises of data on Google playstore applications(apps). Two datasets were utilized, the first dataset, playstore_apps contains information of the applications (apps) and the second, playstore_reviews contains information about the reviews of the apps. 
This only contains the preprocessing of the datasets required for the analysis. Business questions were answered using Structured Query Language(SQL)

In [1]:
# Necessary imports 
import pandas as pd
import numpy as np

**1. Pre-processing of apps dataset**

In [2]:
# Read first dataset into a pandas dataframe
app_data = pd.read_csv("playstore_apps.csv", index_col = 'App')

# Preview dataset and features
app_data.info()

app_data.head(6)

<class 'pandas.core.frame.DataFrame'>
Index: 10841 entries, Photo Editor & Candy Camera & Grid & ScrapBook to iHoroscope - 2018 Daily Horoscope & Astrology
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Category        10841 non-null  object 
 1   Rating          9367 non-null   float64
 2   Reviews         10840 non-null  float64
 3   Size            10841 non-null  object 
 4   Installs        10840 non-null  float64
 5   Type            10840 non-null  object 
 6   Price           10840 non-null  float64
 7   Content Rating  10840 non-null  object 
 8   Genres          10841 non-null  object 
 9   Last Updated    10840 non-null  object 
 10  Current Ver     10833 non-null  object 
 11  Android Ver     10838 non-null  object 
dtypes: float64(4), object(8)
memory usage: 1.1+ MB


Category  Rating  \
App                                                                          
Photo Editor & Candy Camera & Grid & ScrapBook      ART_AND_DESIGN     4.1   
Coloring book moana                                 ART_AND_DESIGN     3.9   
U Launcher Lite – FREE Live Cool Themes, Hide Apps  ART_AND_DESIGN     4.7   
Sketch - Draw & Paint                               ART_AND_DESIGN     4.5   
Pixel Draw - Number Art Coloring Book               ART_AND_DESIGN     4.3   
Paper flowers instructions                          ART_AND_DESIGN     4.4   

                                                     Reviews  Size  \
App                                                                  
Photo Editor & Candy Camera & Grid & ScrapBook         159.0   19M   
Coloring book moana                                    967.0   14M   
U Launcher Lite – FREE Live Cool Themes, Hide Apps   87510.0  8.7M   
Sketch - Draw & Paint                               215644.0   25M   
Pixel Draw - Number Art Coloring Book                  967.0  2.8M   
Paper flowers instructions                             167.0  5.6M   

                                                      Installs  Type  Price  \
App                                                                           
Photo Editor & Candy Camera & Grid & ScrapBook         10000.0  Free    0.0   
Coloring book moana                                   500000.0  Free    0.0   
U Launcher Lite – FREE Live Cool Themes, Hide Apps   5000000.0  Free    0.0   
Sketch - Draw & Paint                               50000000.0  Free    0.0   
Pixel Draw - Number Art Coloring Book                 100000.0  Free    0.0   
Paper flowers instructions                             50000.0  Free    0.0   

                                                   Content Rating  \
App                                                                 
Photo Editor & Candy Camera & Grid & ScrapBook           Everyone   
Coloring book moana                                      Everyone   
U Launcher Lite – FREE Live Cool Themes, Hide Apps       Everyone   
Sketch - Draw & Paint                                        Teen   
Pixel Draw - Number Art Coloring Book                    Everyone   
Paper flowers instructions                               Everyone   

                                                                       Genres  \
App                                                                             
Photo Editor & Candy Camera & Grid & ScrapBook                   Art & Design   
Coloring book moana                                 Art & Design;Pretend Play   
U Launcher Lite – FREE Live Cool Themes, Hide Apps               Art & Design   
Sketch - Draw & Paint                                            Art & Design   
Pixel Draw - Number Art Coloring Book                 Art & Design;Creativity   
Paper flowers instructions                                       Art & Design   

                                                   Last Updated  \
App                                                               
Photo Editor & Candy Camera & Grid & ScrapBook       07-01-2018   
Coloring book moana                                  15-01-2018   
U Launcher Lite – FREE Live Cool Themes, Hide Apps   01-08-2018   
Sketch - Draw & Paint                                08-06-2018   
Pixel Draw - Number Art Coloring Book                20-06-2018   
Paper flowers instructions                           26-03-2017   

                                                           Current Ver  \
App                                                                      
Photo Editor & Candy Camera & Grid & ScrapBook                   1.0.0   
Coloring book moana                                              2.0.0   
U Launcher Lite – FREE Live Cool Themes, Hide Apps               1.2.4   
Sketch - Draw & Paint                               Varies with device   
Pixel Draw - Number Art Coloring Book                   

In [3]:
# Change index of dataset
app_data.reset_index(inplace = True)

**Pre-processing 1.1. Find and remove duplicates**

In [4]:
# Checking for duplicates in the apps dataset
print('Number of duplicates:', app_data.duplicated().sum())

# Removing duplicate data
app_data.drop_duplicates(keep=False, inplace = True)

# Viewing that duplicates have been removed
print('Existing duplicates:', sum(app_data.duplicated()))

Number of duplicates: 483
Existing duplicates: 0


In [5]:
# Overview of dataset after removal of duplicates
app_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9948 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9948 non-null   object 
 1   Category        9948 non-null   object 
 2   Rating          8491 non-null   float64
 3   Reviews         9947 non-null   float64
 4   Size            9948 non-null   object 
 5   Installs        9947 non-null   float64
 6   Type            9947 non-null   object 
 7   Price           9947 non-null   float64
 8   Content Rating  9947 non-null   object 
 9   Genres          9948 non-null   object 
 10  Last Updated    9947 non-null   object 
 11  Current Ver     9940 non-null   object 
 12  Android Ver     9945 non-null   object 
dtypes: float64(4), object(9)
memory usage: 1.1+ MB


In [6]:
# Further examination of the Apps column for duplicates
print('Duplicates:', app_data.App.duplicated().any(), '\n' + 'Number of duplicates:', app_data.App.duplicated().sum())

# Remove duplicates
app_data.drop_duplicates(subset = ['App'], inplace = True)

# Confirm removal
print('Duplicates after cleaning:', app_data.App.duplicated().any())

Duplicates: True 
Number of duplicates: 566
Duplicates after cleaning: False


**Pre-processing 1.2. Find and remove any existing irrelevant value from each column**

In [7]:
# Examine each colum of the apps dataset for irrelevant or unrelated values or entries.

# Examining the unique values in the category column
app_data['Category'].unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

**Observation**: The category column has an irrelevant value entered as "1.9".

In [8]:
# Examining the unique values in the genre column
app_data['Genres'].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Adventure', 'Board', 'Trivia',
       'Simulation', 'Role 

**Observation**: The genre column has an irrelevant value entered as "February 11, 2018". This is a date and cannot be a valid genre.

In [9]:
# Checking rows with category value as 1.9 and genre as February 11, 2018.
app_data.loc[app_data['Genres'] == 'February 11, 2018']

App Category  Rating  Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0      NaN   

         Size  Installs Type  Price Content Rating             Genres  \
10472  1,000+       NaN    0    NaN            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472          NaN  4.0 and up         NaN

In [10]:
# Remove irrelevant value from the genres and category columns.
app_data.drop(10472, inplace = True)

In [11]:
# Again checking if category value as 1.9 and genre as February 11, 2018 removed or not.
app_data.loc[app_data['Genres'] == 'February 11, 2018']

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

In [12]:
# Checking for irrelevant values in the rating column
app_data['Rating'].unique()

array([4.1, 3.9, 4.7, 4.5, 4.3, 4.4, 3.8, 4.2, 4.6, 3.2, 4. , nan, 4.8,
       4.9, 3.6, 3.7, 3.3, 3.4, 3.5, 3.1, 5. , 2.6, 3. , 2.5, 1. , 1.9,
       2.9, 2.8, 2.3, 2.2, 1.7, 2. , 1.8, 2.7, 2.4, 1.6, 2.1, 1.4, 1.5,
       1.2])

**Observation**: There are no irrelevant values in the rating column.

In [13]:
# Checking the unique values and the number of times they appear in the content rating column.
app_data['Content Rating'].value_counts(dropna = False)

Everyone           7695
Teen               1012
Mature 17+          359
Everyone 10+        310
Adults only 18+       3
Unrated               2
Name: Content Rating, dtype: int64

**Observation**: There are no irrelevant values in the content rating column.

In [14]:
# Examining the current ver column for irrelevant values.
app_data['Current Ver'].value_counts(dropna = False).to_frame().head(30)

Current Ver
Varies with device          997
1                           826
1.1                         268
1.2                         182
2                           162
1.3                         138
1.0.0                       132
1.0.1                       112
1.4                          90
1.5                          83
1.0.2                        78
3                            65
1.6                          65
2.1                          63
1.0.3                        60
1.0.4                        55
1.7                          54
2.0.0                        49
1.1.0                        49
1.1.1                        48
1.0.5                        46
1.2.1                        45
1.2.0                        42
1.0.6                        40
4                            40
1.8                          37
1.9                          35
1.0.7                        34
3.1                          33
5                            33

**Observation**: There are no irrelevant values in the current ver column.

In [15]:
# Inspecting the column, "type" for irrelevant values.
app_data['Type'].unique()

array(['Free', 'Paid', nan], dtype=object)

**Observation**: There are no irrelevant values in the type column.

In [16]:
# Inspecting the android ver column for irrelevant values.
app_data['Android Ver'].unique()

array(['4.0.3 and up', '4.2 and up', '4.4 and up', '2.3 and up',
       '3.0 and up', '4.1 and up', '4.0 and up', '2.3.3 and up',
       'Varies with device', '2.2 and up', '5.0 and up', '6.0 and up',
       '1.6 and up', '1.5 and up', '2.1 and up', '7.0 and up',
       '5.1 and up', '4.3 and up', '4.0.3 - 7.1.1', '2.0 and up',
       '3.2 and up', '4.4W and up', '7.1 and up', '7.0 - 7.1.1',
       '8.0 and up', '5.0 - 8.0', '3.1 and up', '2.0.1 and up',
       '4.1 - 7.1.1', nan, '5.0 - 6.0', '1.0 and up', '2.2 - 7.1.1',
       '5.0 - 7.1.1'], dtype=object)

**Observation**: The value, "4.4W and up" appears to have a "W" mistakenly written after the android version number. 

Out of all the entries in the column, this is the only different one.

In [17]:
# Rows with Android Ver '4.4W and up'
app_data.loc[app_data['Android Ver'] == '4.4W and up']

App            Category  Rating  \
1024            Sarajevo Film Festival - Official              EVENTS     NaN   
1089                 WiseBanyan - Invest For Free             FINANCE     4.2   
2289               Critical Care Paramedic Review             MEDICAL     4.4   
5970   Explore British Columbia - BC Travel Guide    TRAVEL_AND_LOCAL     4.4   
6365                              bk Group Mobile            BUSINESS     NaN   
6377                                      SKIN BK        PRODUCTIVITY     NaN   
6444                      BL-ED: From BLUE to RED                GAME     NaN   
8288                          McClatchy DC Bureau  NEWS_AND_MAGAZINES     4.3   
9515                     Sanu Ek Pal Chain - Raid               TOOLS     NaN   
10811                                 FR Plus 1.6   AUTO_AND_VEHICLES     NaN   

       Reviews  Size  Installs  Type  Price Content Rating            Genres  \
1024       6.0  9.7M     100.0  Free   0.00   Everyone 10+            Events   
1089     257.0   44M   10000.0  Free   0.00       Everyone           Finance   
2289      17.0  1.8M    1000.0  Paid   9.99   Everyone 10+           Medical   
5970       7.0   17M     500.0  Free   0.00       Everyone    Travel & Local   
6365       1.0   30M      50.0  Free   0.00       Everyone          Business   
6377       0.0   33M      10.0  Free   0.00       Everyone      Productivity   
6444       3.0  2.9M      10.0  Free   0.00       Everyone            Arcade   
8288      80.0  8.7M    5000.0  Free   0.00   Everyone 10+  News & Magazines   
9515       1.0  2.6M     500.0  Free   0.00       Everyone             Tools   
10811      4.0  3.9M     100.0  Free   0.00       Everyone   Auto & Vehicles   

      Last Updated Current Ver  Android Ver  
1024    31-07-2018       2.0.1  4.4W and up  
1089    31-07-2018      2.2.12  4.4W and up  
2289    14-05-2018       2.0.5  4.4W and up  
5970    09-10-2017       2.0.0  4.4W and up  
6365    26-02-2018           2  4.4W and up  
6377    25-06-2018       1.1.2  4.4W and up  
6444    24-04-2016       1.6.5  4.4W and up  
8288    18-06-2018       6.0.1  4.4W and up  
9515    12-05-2018           2  4.4W and up  
10811   24-07-2018       1.3.6  4.4W and up

In [18]:
# Remove W from the Android Ver column
app_data['Android Ver'] = app_data['Android Ver'].str.replace('W', ' ')

# Checking the rows with Android Ver '4.4W and up' still present or not
app_data.loc[app_data['Android Ver'] == '4.4W and up']

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

In [19]:
# Inspecting for irrelevant values and characters in the app column
app_data.App.sort_values(ascending= True).head(40)

8884                  "i DT" Fútbol. Todos Somos Técnicos.
324                                                 #NAME?
8532                         +Download 4 Instagram Twitter
4541                                                    .R
4636                                                /u/app
5940                                                058.ba
10252                                       1. FC Köln App
8483                         10 Minutes a Day Times Tables
7738        10 WPM Amateur ham radio CW Morse code trainer
8219                            10,000 Quotes DB (Premium)
9401                                  100 Doors of Revenge
4030                                       100+ C Programs
8703       100000+ Messages - DP, Status, Jokes & GIF 2018
4026                            101 C Programming Problems
220      104 Looking for a job - looking for a job, loo...
2694                                                  11st
5344     12 Step Meditations & Sober Prayers AA NA AL-AN

**Observation**: Special characters between strings.

In [20]:
# Create a list of special characters
spec_chars = ["!",'"',"#","%","&","'","(",")",
              "*","+",",","-",".","/",":",";","<",
              "=",">","?","@","[","\\","]","^","_",
              "`","{","|","}","~", "–", "¡", "¿"]


# Remove special characters from App and Category column. But we replaced the special characters with a plus, we might end up with double whitespaces between some words.
for char in spec_chars:
    app_data.App = app_data.App.str.replace(char, ' ', regex=True)
    app_data.Category = app_data.Category.str.replace(char, ' ', regex=True)


# Remove extra whitespaces by splitting each word using whitespaces and re-joining the words again using join.
app_data.App = app_data.App.str.split().str.join(" ")
app_data.Category = app_data.Category.str.split().str.join(" ")

In [21]:
# Examining the app and category column after the removal of special characters.

app_data.sort_values(by = ['App', 'Category'], ascending= (True, False), na_position = 'first').head(30)

App             Category  \
3750                                                       NEWS AND MAGAZINES   
9308                                                       NEWS AND MAGAZINES   
5940                                              058 ba   NEWS AND MAGAZINES   
10252                                      1 FC Köln App               SPORTS   
8219                            10 000 Quotes DB Premium  BOOKS AND REFERENCE   
8483                       10 Minutes a Day Times Tables               FAMILY   
7738      10 WPM Amateur ham radio CW Morse code trainer        COMMUNICATION   
4030                                      100 C Programs               FAMILY   
9401                                100 Doors of Revenge               FAMILY   
8703            100000 Messages DP Status Jokes GIF 2018            LIFESTYLE   
4026                          101 C Programming Problems               FAMILY   
220    104 Looking for a job looking for a job lookin...             BUSINESS   
2694                                                11st             SHOPPING   
5344     12 Step Meditations Sober Prayers AA NA AL ANON            LIFESTYLE   
2483                                       14thStreetVet              MEDICAL   
6760                            17th Edition Cable Sizer  BOOKS AND REFERENCE   
1997                       1LINE One Line with One Touch                 GAME   
7382                           1st Fed CI Mobile Banking              FINANCE   
7993   2 Amateur ham radio CW Morse code practice key...        COMMUNICATION   
7666                         2 Player Co op Zombie Shoot                 GAME   
7338                                       20 Minuten CH   NEWS AND MAGAZINES   
7330                                       20 minutes CH   NEWS AND MAGAZINES   
4959                      2000 AD Comics and Judge Dredd               COMICS   
6518                         2017 BN SM Sales Conference             BUSINESS   
3448       2018Emoji Keyboard Emoticons Lite sticker gif      PERSONALIZATION   
9809                        2019 Tricks Es File Explores                TOOLS   
5205                                             2048 AI               FAMILY   
4272                    2048 BTS J Hope KPop Puzzle Game               FAMILY   
8004   20WPM Amateur ham radio Koch CW Morse code tra...        COMMUNICATION   
7466                                              211 CK                 GAME   

       Rating   Reviews                Size    Installs  Type  Price  \
3750      4.5   44274.0                8.0M   1000000.0  Free   0.00   
9308      4.8     311.0                 14M      5000.0  Free   0.00   
5940      4.4      27.0                 14M       100.0  Free   0.00   
10252     4.6    2019.0                 41M    100000.0  Free   0.00   
8219      4.1      70.0                3.5M       500.0  Paid   0.99   
8483      4.1     681.0                 48M    100000.0  Free   0.00   
7738      3.5      10.0                3.8M       100.0  Paid   1.49   
4030      4.2      20.0                1.6M      5000.0  Free   0.00   
9401      4.1  105766.0                 48M  10000000.0  Free   0.00   
8703      3.7     121.0                3.8M     10000.0  Free   0.00   
4026      4.6     498.0                5.0M     50000.0  Free   0.00   
220       4.4   74359.0                 25M   1000000.0  Free   0.00   
2694      3.8   48732.0                 20M  10000000.0  Free   0.00   
5344      4.7     759.0                 15M     50000.0  Free   0.00   
2483      NaN       0.0                 29M         5.0  Free   0.00   
6760      4.4      47.0                1.4M      1000.0  Paid   3.08   
1997      4.6  214878.0                 10M  10000000.0  Free   0.00   
7382      NaN       0.0                 31M       500.0  Free   0.00   
7993      4.8       6.0                3.5M       100.0  Paid   1.49   
7666      4.2      33.0                 13M      1000.0  Free   0.00   
7338      3.5   14153.0  Vari

**Pre-processing 1.3. Handling null values in the different columns**

In [22]:
# Changing null values to zero in the Rating column

# Checking number of null values
print('Number of null values in the rating column:', app_data['Rating'].isnull().sum())

# Convert Null values to 0 in Rating column as rating is of float data-type
app_data.Rating = app_data.Rating.fillna(0)

# Confirming there are no more null values
print('Number of null values after replacement:', app_data['Rating'].isnull().sum())

Number of null values in the rating column: 1456
Number of null values after replacement: 0


In [23]:
# Preview dataset and features
app_data.info()

print('\n')

# Inspecting the actual number of null values in each column
print('Number of null values in each column:','\n', app_data.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9381 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             9381 non-null   object 
 1   Category        9381 non-null   object 
 2   Rating          9381 non-null   float64
 3   Reviews         9381 non-null   float64
 4   Size            9381 non-null   object 
 5   Installs        9381 non-null   float64
 6   Type            9380 non-null   object 
 7   Price           9381 non-null   float64
 8   Content Rating  9381 non-null   object 
 9   Genres          9381 non-null   object 
 10  Last Updated    9381 non-null   object 
 11  Current Ver     9373 non-null   object 
 12  Android Ver     9379 non-null   object 
dtypes: float64(4), object(9)
memory usage: 1.0+ MB


Number of null values in each column: 
 App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Typ

In [24]:
# Remove rows with a null values in columns with datatype text or string
app_data = app_data.dropna(axis = 0, subset=['Type', 'Android Ver'])

In [25]:
# Changing null values to NaN in Current Ver column.
app_data["Current Ver"].fillna("NaN", inplace = True)

In [26]:
# Examining the number of null values in the columns
app_data.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

**Pre-processing 1.4. Changing incorrect datatypes**

In [27]:
# Change datatype of columns with incorrect datatype
# Column with the incorrect datatype include: Last Updated

app_data['Last Updated'] = pd.to_datetime(app_data['Last Updated'], dayfirst = True)

In [28]:
# Confirming change of datatypes
app_data.dtypes

App                       object
Category                  object
Rating                   float64
Reviews                  float64
Size                      object
Installs                 float64
Type                      object
Price                    float64
Content Rating            object
Genres                    object
Last Updated      datetime64[ns]
Current Ver               object
Android Ver               object
dtype: object

**Pre-processing 1.5. Changing text case in appropriate columns**

In [29]:
# Change the case of text in the category column from upper to proper.
app_data['Category'] = app_data['Category'].str.title()

# Confirm changes have been made
app_data['Category']

0             Art And Design
1             Art And Design
2             Art And Design
3             Art And Design
4             Art And Design
                ...         
10836                 Family
10837                 Family
10838                Medical
10839    Books And Reference
10840              Lifestyle
Name: Category, Length: 9378, dtype: object

In [30]:
# Details of dataset after cleaning, number of rows and columns.
app_data.shape

(9378, 13)

In [31]:
# Save apps dataset after cleaning
app_data.to_csv("playstore_apps_clean_team_datacrawlers.csv", encoding= 'utf-8')

**2. Pre-processing of reviews dataset**

In [32]:
# Read the reviews dataset
review_data = pd.read_csv("playstore_reviews.csv")

# Preview features of dataset
review_data.info()

review_data.sample(6)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


App  \
5794                      All Mental disorders   
44814                         Fashion in Vogue   
111                      10 Best Foods for You   
58748                      HESI A2 Pocket Prep   
53683  Girly Lock Screen Wallpaper with Quotes   
60965                               Helix Jump   

                                       Translated_Review Sentiment  \
5794   This good reading! I'm interested fascinating ...  Positive   
44814  Best coming girl spends 3 hours trying clothes...  Positive   
111    Useful information The amount spelling errors ...  Positive   
58748  great, I like get explanation question get wrong.  Positive   
53683                                                NaN       NaN   
60965  Well, sensitivity option would good, seriously...  Positive   

       Sentiment_Polarity  Sentiment_Subjectivity  
5794             0.398214                0.507143  
44814            1.000000                0.300000  
111              0.200000                0.100000  
58748            0.150000                0.825000  
53683                 NaN                     NaN  
60965            0.183333                0.633333

**Pre-processing 2.1. Finding and removing nulls in the Translated review column**

In [33]:
# Removing rows with NaN reviews in reference to Translated_Review column

# Checking number of null values
print('Number of NaN reviews in the Translated_Review column:', review_data['Translated_Review'].isnull().sum())

# Remove rows with NaN reviews in reference to Translated_Review column
review_data.dropna(subset=['Translated_Review'], inplace= True)

# Check that null values have been removed from the Translated_Review column
review_data['Translated_Review'].isna().any()

Number of NaN reviews in the Translated_Review column: 26868


False

In [34]:
# Preview dataset
review_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 37427 entries, 0 to 64230
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     37427 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37427 non-null  object 
 3   Sentiment_Polarity      37427 non-null  float64
 4   Sentiment_Subjectivity  37427 non-null  float64
dtypes: float64(2), object(3)
memory usage: 1.7+ MB


**Pre-processing 2.2. Finding and removing duplicates**

In [35]:
# Number of duplicate rows in the Translated_Review column
print('Number of duplicates in the Translated_Review column:', review_data.duplicated(subset = ['Translated_Review']).sum())

# Remove duplicates from the reviews dataset
review_data.drop_duplicates(subset = ['Translated_Review'], inplace=True)

# Confirming whether duplicates have been removed or not
print('NUmber of duplicates after cleaning:', review_data.duplicated().sum())

Number of duplicates in the Translated_Review column: 9433
NUmber of duplicates after cleaning: 0


In [36]:
# Preview of dataset after the removal of duplicates
review_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27994 entries, 0 to 64230
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     27994 non-null  object 
 1   Translated_Review       27994 non-null  object 
 2   Sentiment               27994 non-null  object 
 3   Sentiment_Polarity      27994 non-null  float64
 4   Sentiment_Subjectivity  27994 non-null  float64
dtypes: float64(2), object(3)
memory usage: 1.3+ MB


**Pre-processing 2.3. Removing special characters from necessary columns**

In [37]:
# Create a list of special characters
spec_chars_in_reviews = ["!", '"', "#", "%", "&", "(", ")", "*", "+", ",", "-", ".", "/", "é",
              ":", ";", "<", "=", ">", "?", "@", "[", "\\", "]", "^", "_", "`", "⚽", "'", "’",
              "{", "|", "}", "~", "–", "¡", "¿", "❤", "😂", "—", "☹️", "✋", "$", "☝", "'", "喔",
              "♥️", "✌", "🏆", "✔️", "»", "★", "🔥", "လေတွေ ပေးစေလိုပါတယ်", "♡", "ပေမယ့်", "🏠", "'"
              "င့်", "သိချင်တာလေးတွေ", "5️⃣", "¢", "²", "⭐", "♥", "®", "☺", "オリジナル漫画が全話無料で読み放題",
              "蘋果動新聞", "無料レシピ動画で料理を楽しく・簡単に！", "다음 메일", "Доставка еды пицца суши бургер салат",
              "房屋交易 香港", "搵楼租楼", "à", "☹", "☆", "č", "‘", "™", "ë", "ô", "ý", "ß", "し", "â", "×"]

In [38]:
# Remove special characters from App and Translated_Review column. But we replaced the special characters with a plus, we might end up with double whitespaces between some words.
for char in spec_chars_in_reviews:
    review_data.App = review_data.App.str.replace(char, ' ', regex=True)
    review_data.Translated_Review = review_data.Translated_Review.str.replace(char, ' ', regex=True)


# Remove extra whitespaces by splitting each word using whitespaces and re-joining the words again using join.
review_data.App = review_data.App.str.split().str.join(" ")
review_data.Translated_Review = review_data.Translated_Review.str.split().str.join(" ")

In [39]:
# Remove rows with the App names making no sense to ensure utf-8 encoding
review_data = review_data[review_data.App != "104 找工作 找工作 找打工 找兼職 履歷健檢 履歷診療室"]
review_data = review_data[review_data.App != "BÁO MỚI Đọc Báo Tin Tức 24h"]
review_data = review_data[review_data.App != "591房屋交易 租屋、中古屋、新建案、實價登錄、別墅透天、公寓套房、捷運、買房賣房行情、房價房貸查詢"]

In [40]:
# Preview of dataframe after removing special characters.
review_data.head(30)

App                                  Translated_Review  \
0   10 Best Foods for You  I like eat delicious food That s I m cooking f...   
1   10 Best Foods for You    This help eating healthy exercise regular basis   
3   10 Best Foods for You         Works great especially going grocery store   
4   10 Best Foods for You                                       Best idea us   
5   10 Best Foods for You                                           Best way   
6   10 Best Foods for You                                            Amazing   
8   10 Best Foods for You                                Looking forward app   
9   10 Best Foods for You                  It helpful site It help foods get   
10  10 Best Foods for You                                           good you   
11  10 Best Foods for You  Useful information The amount spelling errors ...   
12  10 Best Foods for You  Thank you Great app Add arthritis eyes immunit...   
13  10 Best Foods for You  Greatest ever Completely awesome maintain heal...   
14  10 Best Foods for You             Good health Good health first priority   
16  10 Best Foods for You      Health It s important world either life think   
17  10 Best Foods for You  Mrs sunita bhati I thankful developers to make...   
18  10 Best Foods for You  Very Useful in diabetes age 30 I need control ...   
19  10 Best Foods for You                                  One greatest apps   
20  10 Best Foods for You                                          good nice   
21  10 Best Foods for You                              Healthy Really helped   
22  10 Best Foods for You                                         God health   
23  10 Best Foods for You      HEALTH SHOULD ALWAYS BE TOP PRIORITY ON MYSG5   
24  10 Best Foods for You                              An excellent A useful   
25  10 Best Foods for You                     I found lot wealth form health   
26  10 Best Foods for You                          Because I found important   
27  10 Best Foods for You                                     Healthy Eating   
28  10 Best Foods for You                              Very good Simply good   
29  10 Best Foods for You                                            On test   
30  10 Best Foods for You                                               Good   
31  10 Best Foods for You  Thanks advice Downloaded Adobe reader still ca...   
32  10 Best Foods for You                  No recipe book Unable recipe book   

   Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0   Positive            1.000000                0.533333  
1   Positive            0.250000                0.288462  
3   Positive            0.400000                0.875000  
4   Positive            1.000000                0.300000  
5   Positive            1.000000                0.300000  
6   Positive            0.600000                0.900000  
8    Neutral            0.000000                0.000000  
9    Neutral            0.000000                0.000000  
10  Positive            0.700000                0.600000  
11  Positive            0.200000                0.100000  
12  Positive            0.750000                0.875000  
13  Positive            0.992188                0.866667  
14  Positive            0.550000                0.511111  
16  Positive            0.450000                1.000000  
17  Positive            0.600000                0.666667  
18  Positive            0.295000                0.100000  
19  Positive            1.000000                1.000000  
20  Positive            0.650000                0.800000  
21  Positive            0.350000                0.350000  
22   Neutral            0.000000                0.000000  
23  Positive            0.781250                0.500000  
24  Positive            0.650000                0.500000  
25   Neutral            0.000000                0.000000  
26  Positive            0.400000                1.000000  
27  Positive            0.500000                0.500000  
28  Pos

In [41]:
review_data.reset_index(drop = True, inplace = True)

In [42]:
# Total rows in reviews dataset after cleaning

review_data.shape

(27914, 5)

In [43]:
# Save reviews dataset after cleaning to a csv file
review_data.to_csv('playstore_reviews_clean_teamdatacrawlers.csv', encoding='utf-8')

**CLEANING DOCUMENTATION**

**◘ Apps Dataset**
1. The playstore apps dataset contains 13 columns and 10841 rows. And there were 483 duplicates in the dataset that were removed. There were also 566 duplicates present in the 'App' column that were removed.
2. Irrelevant values were also removed if present in any column, for instance '1.9' was present in 'Category' column.
3. Special characters like '+', '_', '-' and more were also removed from the 'App' and 'Category' column.
4. Null values in the 'Rating' column were changed to zero.
5. Rows with null values in columns having datatype text or string were also removed.
6. Null values present in the 'Current Ver' column were changed to NaN.
7. The 'Last Updated' column contains dates. Its datatype was changed from object to datetime to reflect the observations within.
8. Generated csv file of playstore apps dataset after cleaning.

**◘ Reviews Dataset** 
1. The playstore reviews dataset contains 5 columns and 64295 rows. And there were 9433 duplicates in the 'Translated_Review' column that were removed.
2. Rows with NaN reviews in the 'Translated_Reviews' column of the playstore reviews dataset were dropped since they won't provide any meaningful information.
3. Removed special characters from 'App' and 'Translated_Review' column in order to ensure utf-8 encoding